File: app/fsharp/src/Ch03/ImmutableOperations.fs
namespace Ch03

/// 第3章: イミュータブルなデータ操作
/// リストのスライスと結合操作を学ぶ
module ImmutableOperations =

    // ============================================
    // 基本的なリスト操作
    // ============================================

    /// リストの先頭n個の要素を取得
    let firstN (n: int) (list: 'a list) : 'a list =
        list |> List.truncate n

    /// リストの末尾n個の要素を取得
    let lastN (n: int) (list: 'a list) : 'a list =
        let skipCount = max 0 (List.length list - n)
        list |> List.skip skipCount

    /// リストの最初の2要素を取得
    let firstTwo (list: 'a list) : 'a list =
        firstN 2 list

    /// リストの最後の2要素を取得
    let lastTwo (list: 'a list) : 'a list =
        lastN 2 list

    /// リストのスライス（開始インデックスから終了インデックスまで）
    let slice (startIndex: int) (endIndex: int) (list: 'a list) : 'a list =
        list
        |> List.skip startIndex
        |> List.truncate (endIndex - startIndex)

    // ============================================
    // 要素の追加
    // ============================================

    /// リストの末尾に要素を追加（新しいリストを返す）
    let appended (element: 'a) (list: 'a list) : 'a list =
        list @ [element]

    /// リストの末尾に複数の要素を追加
    let appendedAll (elements: 'a list) (list: 'a list) : 'a list =
        list @ elements

    // ============================================
    // リスト変換パターン
    // ============================================

    /// 最初の2要素を末尾に移動
    let moveFirstTwoToEnd (list: 'a list) : 'a list =
        let first = firstTwo list
        let rest = list |> List.skip 2
        rest @ first

    /// 最後の要素の前に新しい要素を挿入
    let insertBeforeLast (element: 'a) (list: 'a list) : 'a list =
        let withoutLast = list |> List.truncate (List.length list - 1)
        let last = list |> lastN 1
        withoutLast @ [element] @ last

    /// 中央に要素を挿入
    let insertAtMiddle (element: 'a) (list: 'a list) : 'a list =
        let middle = List.length list / 2
        let before = list |> List.truncate middle
        let after = list |> List.skip middle
        before @ [element] @ after


/// 旅程の再計画
module Itinerary =

    /// 指定した都市の前に新しい都市を挿入
    let replan (plan: string list) (newCity: string) (beforeCity: string) : string list =
        let index =
            plan
            |> List.tryFindIndex (fun c -> c = beforeCity)
            |> Option.defaultValue (List.length plan)
        let citiesBefore = plan |> List.truncate index
        let citiesAfter = plan |> List.skip index
        citiesBefore @ [newCity] @ citiesAfter


/// 文字列操作
module StringOperations =

    /// 名前を省略形に変換（例: "Alonzo Church" -> "A. Church"）
    let abbreviate (name: string) : string =
        let initial = name.Substring(0, 1)
        let separatorIndex = name.IndexOf(' ')
        if separatorIndex < 0 then
            name
        else
            let lastName = name.Substring(separatorIndex + 1)
            $"{initial}. {lastName}"

    /// 文字列の最初のn文字を取得
    let firstNChars (n: int) (s: string) : string =
        if n >= s.Length then s
        else s.Substring(0, n)

    /// 文字列の最後のn文字を取得
    let lastNChars (n: int) (s: string) : string =
        if n >= s.Length then s
        else s.Substring(s.Length - n)

File: app/csharp/src/Ch03/ImmutableOperations.cs
using LanguageExt;
using static LanguageExt.Prelude;

namespace Ch03;

/// <summary>
/// 第3章: イミュータブルなデータ操作
/// リストのスライスと結合操作を学ぶ
/// </summary>
public static class ImmutableOperations
{
    // ============================================
    // 基本的なリスト操作
    // ============================================

    /// <summary>
    /// Seq の先頭 n 個の要素を取得
    /// </summary>
    public static Seq<T> FirstN<T>(Seq<T> seq, int n) =>
        seq.Take(n);

    /// <summary>
    /// Seq の末尾 n 個の要素を取得
    /// </summary>
    public static Seq<T> LastN<T>(Seq<T> seq, int n)
    {
        var skipCount = Math.Max(0, seq.Count - n);
        return seq.Skip(skipCount);
    }

    /// <summary>
    /// Seq の最初の 2 要素を取得
    /// </summary>
    public static Seq<T> FirstTwo<T>(Seq<T> seq) =>
        FirstN(seq, 2);

    /// <summary>
    /// Seq の最後の 2 要素を取得
    /// </summary>
    public static Seq<T> LastTwo<T>(Seq<T> seq) =>
        LastN(seq, 2);

    /// <summary>
    /// Seq のスライス（開始インデックスから終了インデックスまで）
    /// </summary>
    public static Seq<T> Slice<T>(Seq<T> seq, int startIndex, int endIndex) =>
        seq.Skip(startIndex).Take(endIndex - startIndex);

    // ============================================
    // 要素の追加
    // ============================================

    /// <summary>
    /// Seq の末尾に要素を追加（新しい Seq を返す）
    /// </summary>
    public static Seq<T> Appended<T>(Seq<T> seq, T element) =>
        seq.Add(element);

    /// <summary>
    /// Seq の末尾に複数の要素を追加
    /// </summary>
    public static Seq<T> AppendedAll<T>(Seq<T> seq, Seq<T> elements) =>
        seq.Concat(elements);

    // ============================================
    // リスト変換パターン
    // ============================================

    /// <summary>
    /// 最初の 2 要素を末尾に移動
    /// </summary>
    public static Seq<T> MoveFirstTwoToEnd<T>(Seq<T> seq)
    {
        var first = FirstTwo(seq);
        var rest = seq.Skip(2);
        return rest.Concat(first);
    }

    /// <summary>
    /// 最後の要素の前に新しい要素を挿入
    /// </summary>
    public static Seq<T> InsertBeforeLast<T>(Seq<T> seq, T element)
    {
        var withoutLast = seq.Take(seq.Count - 1);
        var last = LastN(seq, 1);
        return withoutLast.Add(element).Concat(last);
    }

    /// <summary>
    /// 中央に要素を挿入
    /// </summary>
    public static Seq<T> InsertAtMiddle<T>(Seq<T> seq, T element)
    {
        var middle = seq.Count / 2;
        var before = seq.Take(middle);
        var after = seq.Skip(middle);
        return before.Add(element).Concat(after);
    }
}

/// <summary>
/// 旅程の再計画
/// </summary>
public static class Itinerary
{
    /// <summary>
    /// 指定した都市の前に新しい都市を挿入
    /// </summary>
    public static Seq<string> Replan(Seq<string> plan, string newCity, string beforeCity)
    {
        var index = plan.ToList().FindIndex(c => c == beforeCity);
        if (index < 0) index = plan.Count;

        var citiesBefore = plan.Take(index);
        var citiesAfter = plan.Skip(index);
        return citiesBefore.Add(newCity).Concat(citiesAfter);
    }
}

/// <summary>
/// 文字列操作
/// </summary>
public static class StringOperations
{
    /// <summary>
    /// 名前を省略形に変換（例: "Alonzo Church" -> "A. Church"）
    /// </summary>
    public static string Abbreviate(string name)
    {
        var initial = name.Substring(0, 1);
        var separatorIndex = name.IndexOf(' ');
        if (separatorIndex < 0)
            return name;

        var lastName = name.Substring(separatorIndex + 1);
        return $"{initial}. {lastName}";
    }

    /// <summary>
    /// 文字列の最初の n 文字を取得
    /// </summary>
    public static string FirstNChars(string s, int n) =>
        n >= s.Length ? s : s.Substring(0, n);

    /// <summary>
    /// 文字列の最後の n 文字を取得
    /// </summary>
    public static string LastNChars(string s, int n) =>
        n >= s.Length ? s : s.Substring(s.Length - n);
}
